DDG4/edm4hep/EDM4hepFileReader.cpp

```diff
@@ -0,0 +1,332 @@
+
+#include <podio/Frame.h>
+#include <podio/Reader.h>
+
+typedef dd4hep::detail::ReferenceBitMask<int> PropertyMask;
+
+
+namespace dd4hep::sim {
+
+  // we use the index of the objectID to identify particles
+  // we will not support MCParticles from different collections
+  using MCPARTICLE_MAP=std::map<int, int>;
+
+  /// get the parameters from the GenericParameters of the input EDM4hep frame and store them in the EventParameters
+  /// extension
+  template <class T=podio::GenericParameters> void EventParameters::ingestParameters(T const& source) {
+    for(auto const& key: source.template getKeys<int>()) {
+      m_intValues[key] = source.template get<std::vector<int>>(key).value();
+    }
+    for(auto const& key: source.template getKeys<float>()) {
+      m_fltValues[key] = source.template get<std::vector<float>>(key).value();
+    }
+    for(auto const& key: source.template getKeys<double>()) {
+      m_dblValues[key] = source.template get<std::vector<double>>(key).value();
+    }
+    for(auto const& key: source.template getKeys<std::string>()) {
+      m_strValues[key] = source.template get<std::vector<std::string>>(key).value();
+    }
+  }
+
+  /// get the parameters from the GenericParameters of the input EDM4hep run frame and store them in the RunParameters
+  /// extension
+  template <class T=podio::GenericParameters> void RunParameters::ingestParameters(T const& source) {
+    for(auto const& key: source.template getKeys<int>()) {
+      m_intValues[key] = source.template get<std::vector<int>>(key).value();
+    }
+    for(auto const& key: source.template getKeys<float>()) {
+      m_fltValues[key] = source.template get<std::vector<float>>(key).value();
+    }
+    for(auto const& key: source.template getKeys<double>()) {
+      m_dblValues[key] = source.template get<std::vector<double>>(key).value();
+    }
+    for(auto const& key: source.template getKeys<std::string>()) {
+      m_strValues[key] = source.template get<std::vector<std::string>>(key).value();
+    }
+  }
+
+  template <class T=podio::GenericParameters> void FileParameters::ingestParameters(T const& source) {
+    for(auto const& key: source.template getKeys<int>()) {
+      m_intValues[key] = source.template get<std::vector<int>>(key).value();
+    }
+    for(auto const& key: source.template getKeys<float>()) {
+      m_fltValues[key] = source.template get<std::vector<float>>(key).value();
+    }
+    for(auto const& key: source.template getKeys<double>()) {
+      m_dblValues[key] = source.template get<std::vector<double>>(key).value();
+    }
+    for(auto const& key: source.template getKeys<std::string>()) {
+      m_strValues[key] = source.template get<std::vector<std::string>>(key).value();
+    }
+  }
+
+  /// Class to read EDM4hep files
+  /**
+   *  \version 1.0
+   *  \ingroup DD4HEP_SIMULATION
+   */
+  class EDM4hepFileReader : public Geant4EventReader {
+  protected:
+    /// Reference to reader object
+    podio::Reader m_reader;
```

Comment:
Looking at CI results, suggesting to use the `Reader` interface here was not entirely sensible, as that would mean we need at least podio 1.0. Sorry, I didn't think that one entirely through.

I am not sure whether we should simply bump the minimum podio version to 1.0. I suppose that is a bit easier to accommodate for EIC than a bump in EDM4hep minimum version @wdconinc ?

Otherwise we would either need to go to quite some pre-processor gymnastics or we could dial this back to only read the ROOT format for now (which would require only a few changes, as far as I can judge right now).

Reply:
I guess I'll be pinged on every backwards compatibility decision now :-)

We're already using podio-1.1 now (but edm4hep-0.10.5), so we'd be able to accommodated a minimum podio version of 1.0.

Reply:
Well, at least the ones where I am aware of a potential issue ;) Also once we are in post 1.0 land a breaking change has a much clearer association with a version number than it does below that, so I hope this will be among the last times we have to ask.

Thanks for the confirmation.

Reply:
Lets enable the reader, when podio>1 ?